src/main/java/org/truffleruby/parser/lexer/StringTerm.java

```diff
@@ -283,6 +285,13 @@ public int parseStringIntoBuffer(RubyLexer lexer, RopeBuilder buffer, Encoding e
                             break;
                         }
                         if (expand) {
+                            if (!(indent || lexer.getHeredocIndent() >= 0)) continue;
+                            if (c == end) {
+                                c = '\\';
+                                // goto terminate
+                                if (enc != null) buffer.setEncoding(lexer.getEncoding());
+                                return c;
+                            }
```

Comment:
To me, this seems like the code snippet most responsible for fixing the bug, along with the check on `\\` in `HeredocTerm`. I'm not sure if any other change than these two are necessary.

I also suspect this does not cause the new issue, which does not involve `\\`.

Reply:
Sweet!!! When I was investigating, I found that removing the change in this code snippet causes the original test to fail too but I wasn't sure why. 

This is great - thanks for taking the time to isolate the actual bug fix. 

cc: @tomstuart